### PR TITLE
[core] Compare placement group objects by ID

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Placement group objects are currently compared by memory address. When sending a placement group reference via the object store, it may get dereferenced in a different object. This can then fail subsequent operations, e.g. if placement groups are compared or tracked in a set.

This PR adds the `__eq__` and `__hash__` methods to the `PlacementGroup` class, which will consider placement group objects as equal when they have the same placement group ID.

Previously failing test snippet:

```
    from copy import copy

    pg1 = ray.util.placement_group([{"CPU": 1}])
    pg2 = copy(pg1)

    # __eq__
    assert pg1 == pg2

    # __hash__
    s = set()
    s.add(pg1)
    assert pg2 in s
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_placement_group.py [hunk 1]
```diff
@@ -512,6 +512,21 @@ def test_placement_group_empty_bundle_error(ray_start_regular, connect_to_client
             ray.util.placement_group([])
 
 
+def test_placement_group_equal_hash(ray_start_regular):
+    from copy import copy
+
+    pg1 = ray.util.placement_group([{"CPU": 1}])
+    pg2 = copy(pg1)
```
Comment (rkooo567) [inline python/ray/tests/test_placement_group.py:519]: Should we test with `ray.put(pg1); p = ray.get(pg1); assert  p == pg1?

**Review by rkooo567 — APPROVED** [review]
It makes sense. Btw, is it hypothetical, or did you actually have the use case of ray.put pg?

**Comment by krfricke** [comment]
We came across it when trying to use a shared Ray AIR resource manager: https://discuss.ray.io/t/shared-resource-manager/9638/4?u=zhang_david

I think the current test is enough but it doesn't hurt to add the remote task case, so I'll update the PR

**Review by rkooo567 — APPROVED** [review]
LGTM

**Review by jjyao — APPROVED** [review]

**Pull request merged**